src/login_nopam.c

```diff
@@ -149,33 +149,35 @@ list_match(char *list, const char *item, bool (*match_fn)(char *, const char*))
 	static const char  sep[] = ", \t";
 
 	char *tok;
+	bool inclusion = true;
+	bool matched = false;
+	bool result = false;
 
 	/*
 	 * Process tokens one at a time. We have exhausted all possible matches
 	 * when we reach an "EXCEPT" token or the end of the list. If we do find
-	 * a match, look for an "EXCEPT" list and recurse to determine whether
-	 * the match is affected by any exceptions.
+	 * a match, look for an "EXCEPT" list and determine whether the match is
+	 * affected by any exceptions.
 	 */
 	while (NULL != (tok = strsep(&list, sep))) {
-		if (strcasecmp (tok, "EXCEPT") == 0) {	/* EXCEPT: give up */
-			break;
+		if (strcasecmp (tok, "EXCEPT") == 0) {	/* EXCEPT: invert */
+			if (!matched) {	/* stop processing: not part of list */
+				break;
+			}
+			inclusion = !inclusion;
+			matched = false;
 		} else {
 			bool  match;
 
 			match = (*match_fn)(tok, item);
 			if (match) {
-				while (   (NULL != (tok = strsep(&list, sep)))
-				       && (strcasecmp (tok, "EXCEPT") != 0))
-					/* VOID */ ;
-				if (tok == NULL || !list_match(list, item, match_fn)) {
-					return (match);
-				}
-				break;
+				result = inclusion;
+				matched = true;
 			}
 		}
 	}
 
-	return false;
+	return result;
```

Comment:
Would you mind explaining a little bit the refactor?  I mean, it was hard to understand how it worked, and thus it's hard to verify that the change is correct.  Since you've done it, you can probably add some words about it (maybe good for the commit message too).

Also, if you find some other temporary refactor would simplify this further, please go ahead!

Reply:
I've added a little bit of text in the linux-pam PR linked above. If it's about the general concept of the code, I guess it's best to keep it there to have a single point of discussion when it comes to correctness.

When it's about shadow-specific style etc. we can do it here

Reply:
Hmmm, I see.  You wrote an algorithm more-or-less from scratch, right?

I'll try to mechanically compare the algorithms with pencil and paper.  :)

Reply:
I've added more context here: https://github.com/linux-pam/linux-pam/pull/871#issuecomment-2597014912